File: src/controllers/userController.py
import time
import asyncio
import logging

logger = logging.getLogger("NP_ULTRA")
USER_CACHE = {}

async def db_syncer(db_conn):
    while True:
        await asyncio.sleep(15)
        if not USER_CACHE or not db_conn: continue
        try:
            cache_copy = USER_CACHE.copy()
            USER_CACHE.clear() 
            to_save = []
            now_ts = int(time.time())
            for uid, entry in cache_copy.items():
                d = entry.get("data", {})
                to_save.append((
                    str(uid), float(d.get('score', 0)), int(d.get('tap_power', 1)), 
                    float(d.get('pnl', 0)), float(d.get('energy', 0)), 
                    int(d.get('max_energy', 1000)), int(d.get('level', 1)), now_ts
                ))
            if to_save:
                async with db_conn.execute("BEGIN TRANSACTION"):
                    await db_conn.executemany("""
                        INSERT INTO users (id, balance, click_lvl, pnl, energy, max_energy, level, last_active)
                        VALUES (?,?,?,?,?,?,?,?) ON CONFLICT(id) DO UPDATE SET
                        balance=excluded.balance, click_lvl=excluded.click_lvl, pnl=excluded.pnl,
                        energy=excluded.energy, level=excluded.level, last_active=excluded.last_active
                    """, to_save)
                await db_conn.commit()
                logger.info(f"💾 Sync OK: {len(to_save)} users.")
        except Exception as e: logger.error(f"Sync Error: {e}")

File: src/controllers/userController.js
const sqlite3 = require('sqlite3');
const { open } = require('sqlite');
const path = require('path');

const dbPath = path.join(__dirname, '../../../data/game.db');

async function getDB() {
    const db = await open({ filename: dbPath, driver: sqlite3.Database });
    await db.exec(`CREATE TABLE IF NOT EXISTS users (
        id TEXT PRIMARY KEY, balance REAL, click_lvl INTEGER, 
        pnl REAL DEFAULT 0, energy REAL, max_energy INTEGER, 
        level INTEGER DEFAULT 1, last_active INTEGER)`);
    return db;
}

exports.getBalance = async (req, res) => {
    const db = await getDB();
    const userId = req.params.userId;
    const now = Math.floor(Date.now() / 1000);

    let user = await db.get('SELECT * FROM users WHERE id = ?', [userId]);

    if (user) {
        const offTime = Math.min(now - (user.last_active || now), 10800);
        const earned = (user.pnl / 3600) * offTime;
        res.json({ status: "ok", data: {
            score: user.balance + earned, tap_power: user.click_lvl,
            pnl: user.pnl, energy: user.energy, level: user.level,
            max_energy: user.max_energy, multiplier: 1
        }});
    } else {
        await db.run(`INSERT INTO users (id, balance, click_lvl, pnl, energy, max_energy, level, last_active) 
                      VALUES (?, 1000, 1, 0, 1000, 1000, 1, ?)`, [userId, now]);
        res.json({ status: "ok", data: { score: 1000, tap_power: 1, pnl: 0, energy: 1000, max_energy: 1000, level: 1, multiplier: 1 }});
    }
};

exports.saveState = async (req, res) => {
    const db = await getDB();
    const d = req.body;
    const now = Math.floor(Date.now() / 1000);

    if (!d.user_id || d.user_id === "guest") return res.json({status: "ignored"});

    await db.run(`UPDATE users SET balance=?, click_lvl=?, pnl=?, energy=?, level=?, last_active=? WHERE id=?`,
        [d.score, d.tap_power, d.pnl, d.energy, d.level, now, d.user_id]);
    
    res.json({status: "ok"});
};
